pkg/export/disk/disk.go

```diff
@@ -0,0 +1,311 @@
+package disk
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/export/util"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/logging"
+	"k8s.io/client-go/util/retry"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+type Connection struct {
+	// path to store audit logs
+	Path string `json:"path,omitempty"`
+	// max number of audit results to store
+	MaxAuditResults int `json:"maxAuditResults,omitempty"`
+	// File to write audit logs
+	File *os.File
+
+	// current audit run file name
+	currentAuditRun string
+}
+
+type Writer struct {
+	openConnections map[string]Connection
+}
+
+const (
+	Name                = "disk"
+	maxAllowedAuditRuns = 5
+	maxAuditResults     = "maxAuditResults"
+	violationPath       = "path"
+)
+
+var Connections = &Writer{
+	openConnections: make(map[string]Connection),
+}
+
+var log = logf.Log.WithName("disk-driver").WithValues(logging.Process, "export")
+
+func (r *Writer) CreateConnection(_ context.Context, connectionName string, config interface{}) error {
+	path, maxResults, err := unmarshalConfig(config)
+	if err != nil {
+		return fmt.Errorf("error creating connection %s: %w", connectionName, err)
+	}
+
+	r.openConnections[connectionName] = Connection{
+		Path:            path,
+		MaxAuditResults: int(maxResults),
+	}
+	return nil
+}
+
+func (r *Writer) UpdateConnection(_ context.Context, connectionName string, config interface{}) error {
+	conn, exists := r.openConnections[connectionName]
+	if !exists {
+		return fmt.Errorf("connection %s for disk driver not found", connectionName)
+	}
+
+	path, maxResults, err := unmarshalConfig(config)
+	if err != nil {
+		return fmt.Errorf("error updating connection %s: %w", connectionName, err)
+	}
+
+	if conn.Path != path {
+		if conn.File != nil {
+			if err := conn.unlockAndCloseFile(); err != nil {
+				return fmt.Errorf("error updating connection %s, error closing file: %w", connectionName, err)
+			}
+		}
+		if err := os.RemoveAll(conn.Path); err != nil {
+			return fmt.Errorf("error updating connection %s, error deleting violations stored at old path: %w", connectionName, err)
+		}
+		conn.Path = path
+		conn.File = nil
+	}
+
+	conn.MaxAuditResults = int(maxResults)
+
+	r.openConnections[connectionName] = conn
+	return nil
+}
+
+func (r *Writer) CloseConnection(connectionName string) error {
+	conn, ok := r.openConnections[connectionName]
+	if !ok {
+		return fmt.Errorf("connection %s not found for disk driver", connectionName)
+	}
+	delete(r.openConnections, connectionName)
```

Comment:
Should we remove the connection only after successfully closing and removing files?